impl/client/src/test/javascript/pentaho/data/filter/abstract.spec.js

```diff
@@ -176,302 +177,475 @@ define([
+        function expectAndNot(specIn, specSubstract, specOut) {
+          var fDnfIn = specToDnf(specIn);
+          var fDnfSubtract = specToDnf(specSubstract);
 
-        expectDnf({
-          _: "and",
-          o: [
-            {
-              _: "or",
-              o: [
-                {_: "=", p: "a", v: 1},
-                {_: "=", p: "b", v: 1}
-              ]
-            },
-            {
-              _: "or",
-              o: [
-                {_: "=", p: "c", v: 1},
-                {_: "=", p: "d", v: 1}
-              ]
-            }
-          ]
-        }, {
-          _: "or",
-          o: [
-            {
-              _: "and",
-              o: [
-                {_: "=", p: "a", v: 1},
-                {_: "=", p: "c", v: 1}
-              ]
-            },
-            {
-              _: "and",
-              o: [
-                {_: "=", p: "a", v: 1},
-                {_: "=", p: "d", v: 1}
-              ]
-            },
-            {
+          var fOut = fDnfIn.andNot(fDnfSubtract);
+
+          if (specOut) {
+            expect(fOut.toSpec({forceType: true})).toEqual(specOut);
+          }
+        }
+
+        it("correctly subtracts isEqual predicates", function () {
+          // (a=1) or ((a=2) and (b=3))
+          var originalSpec = {
+            _: "or",
+            o: [
+              {_: "=", p: "a", v: 1},
+              {
+                _: "and",
+                o: [
+                  {_: "=", p: "a", v: 2},
+                  {_: "=", p: "b", v: 3}
+                ]
+              }
+            ]
+          };
+
+          // a=1
+          var subtractSpec1 = {_: "=", p: "a", v: 1};
+
+          // a=2
+          var subtractSpec2 = {_: "=", p: "a", v: 2};
+
+          // (a=2) and (b=3)
+          var expected1 = {
+            _: "or",
+            o: [{
               _: "and",
               o: [
-                {_: "=", p: "b", v: 1},
-                {_: "=", p: "c", v: 1}
+                {_: "=", p: "a", v: 2},
+                {_: "=", p: "b", v: 3}
               ]
-            },
-            {
+            }]
+          };
+
+          // (a=1)
+          var expected2 = {
+            _: "or",
+            o: [{
               _: "and",
-              o: [
-                {_: "=", p: "b", v: 1},
-                {_: "=", p: "d", v: 1}
-              ]
-            }
-          ]
-        });
-      });
+              o: [{_: "=", p: "a", v: 1}]
+            }]
+          };
 
-      it("should allow subtraction", function() {
+          expectAndNot(originalSpec, subtractSpec1, expected1);
+          expectAndNot(originalSpec, subtractSpec2, expected2);
 
-        expectDnf({
-          _: "and",
-          o: [
-            {_: "=", p: "a", v: 1},
-            {
-              _: "not",
-              o: {
-                _: "or",
+        });
+
+        it("correctly deals with terminal filters other than isEquals", function(){
```

Comment:
I think you are missing a test for the case where it is `b` that has an inequality operator. I also think that that case is currently unhandled in the code.